# Una introducción a la combinación de datos con Pandas `Merge`, `join` y `Concat`

Este cuaderno contiene ejemplos de código para usar con el artículo del mismo nombre, junto con comentarios ligeros sobre cada uno de los ejemplos.

In [1]:
import pandas as pd

pd.set_option("display.max_columns", 50)

## Importar datos

Aquí importará los conjuntos de datos normales del clima de temperatura y precipitación en los marcos de datos. Llamar a `.head()` en el marco de datos le dará una vista previa de 5 filas de sus datos, y el atributo `Shape` le dará las dimensiones de los datos en el formulario `(filas, columnas)`. Estas son excelentes controles de cordura para ejecutarse antes de hacer demasiado con los datos.

In [2]:
climate_temp = pd.read_csv("climate_temp.csv")
climate_temp.head()

STATION            STATION_NAME ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  
0                6                7               10               15  
1                6                7               10               15  
2                5                7               10               15  
3                5                7               10               15  
4                5                7               10               15

In [3]:
climate_temp.shape

(127020, 21)

In [4]:
climate_precip = pd.read_csv("climate_precip.csv")
climate_precip.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6            -6.66             -666   
1             -666            -66.6            -6.66             -666   
2             -666            -66.6            -6.66             -666   
3            -9999          -9999.0            -6.66            -9999   
4            -9999          -9999.0            -6.66            -9999   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6            -6.66             -666            -66.6   
1            -66.6            -6.66             -666            -66.6   
2            -66.6            -6.66             -666            -66.6   
3          -9999.0            -6.66            -9999          -9999.0   
4          -9999.0            -6.66            -9999          -9999.0   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.02              0.0             0.02              0.0   
1             0.04              0.0             0.04              0.0   
2             0.05              0.0             0.05              0.0   
3             0.07              0.0             0.07              0.0   
4             0.09              0.0             0.09              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                       98                       43                       12   
1                       99                       44                       12   
2                      100                       44                       12   
3                      101                       45                       12   
4                      102                       45                       12   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                        3                    -9999                        0   
1                        3                    -9999                        0   
2                        3                    -9999                        0   
3                        3                        0                        0   
4                        3                        0                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                    -9999                    -9999                    -9999   
1                    -9999                    -9999                    -9999   
2                    -9999                    -9999                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                    -9999                        0                    -9999   
1                    -9999                        0                    -9999   
2                    -9999                        0                    -9999   
3                        0                        0                        0   
4                        0                        0                        0   

   DLY-SNOW-PCTALL-GE050TI  
0                    -9999  
1                    -9999  
2                    -9999  
3                        0  
4                        0

In [5]:
climate_precip.shape

(151110, 29)

## merge()

En esta sección, aprenderá sobre la funcionalidad `merge()` en pandas.

### Inner Join

Aquí, usará un simple `merge()` llamar para hacer una unión interna y aprender cómo esto puede resultar en un conjunto de datos más pequeño y más enfocado. Primero, creará un nuevo objeto DataFrame que contiene los datos de precipitación de una estación.

In [6]:
precip_one_station = climate_precip.query("STATION == 'GHCND:USC00045721'")
precip_one_station.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141                       86   

      DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
1460                       31                       11   
1461                       31                       11   
1462                       31                       11   
1463                       32                       11   
1464                       32                       11   

      DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
1460                        4                        0   
1461                        4                        0   
1462                        4                        0   
1463                        4                        0   
1464                        4                        0   

      DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
1460                        3                        3   
1461                        3                        3   
1462                        3                        3   
1463                        3                        2   
1464                        3                        2   

      DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
1460                        9                        6   
1461                       10                        6   
1462                       10                        6   
1463                       10                        6   
1464                       10                        6   

      DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
1460                        0                    -9999   
1461                        0                    -9999   
1462                        0                    -9999   
1463                        0                    -9999   
1464                        0                    -9999   

      DLY-SNOW-PCTALL-GE050TI  
1460                    -9999  
1461              

In [7]:
precip_one_station.shape

(365, 29)

In [8]:
inner_merged = pd.merge(precip_one_station, climate_temp)
inner_merged.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI ELEVATION LATITUDE  LONGITUDE  DLY-CLDD-BASE45  \
0                    -9999    1325.9  34.9436  -115.5469                3   
1                    -9999    1325.9  34.9436  -115.5469                3   
2                    -9999    1325.9  34.9436  -115.5469                3   
3            

¿Cuántas filas crees que tiene este marco de datos fusionado?

In [9]:
inner_merged.shape

(365, 47)

Obtiene 365 filas porque cualquier fila no coincidente se descartan en una unión interna, que es el método de fusión predeterminado para una llamada `merge()` y `precip_one_station` tenía solo 365 filas.

¿Qué pasa si desea fusionar ambos conjuntos de datos completos, pero especifique en qué columnas unirse? En este caso, utilizará el parámetro `on`:

In [10]:
inner_merged_total = pd.merge(
    climate_temp, climate_precip, on=["STATION", "DATE"]
)
inner_merged_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

In [11]:
inner_merged_total.shape

(123005, 48)

Puede especificar una sola columna _Key con una cadena o múltiples columnas de clave con una lista, como en el ejemplo anterior. Esto da como resultado un marco de datos con 123005 filas y 48 columnas.

¿Por qué 48 columnas en lugar de 47? Debido a que especificó las columnas de teclas para unirse, Pandas no intenta fusionar todas las columnas fusionables. Esto puede dar lugar a nombres de columnas "duplicados", que pueden o no tener valores diferentes. "Duplicate" está en citas porque las columnas realmente tendrán nuevos nombres, de forma predeterminada se adjuntan con `_x` y` _y`. También puede usar el parámetro `sufijos` para controlar lo que se adjunta a los nombres de la columna.

### Outer Join
Con la unión externa, retendrá filas que no tienen partidos también. Para este ejemplo, utilizará el DataSframe de datos de precipitación más pequeño `precip_one_station` con las columnas de` climat_temp` completas y se unirán con las columnas `station` y `date` como las columnas clave. Tome un segundo y piense en cuántas filas espera que tenga el nuevo cuadro de datos.

In [12]:
outer_merged = pd.merge(
    precip_one_station, climate_temp, how="outer", on=["STATION", "DATE"]
)
outer_merged.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0           -666.0            -66.6             0.16           -666.0   
1           -666.0            -66.6             0.16           -666.0   
2           -666.0            -66.6             0.16           -666.0   
3           -666.0            -66.6             0.16           -666.0   
4           -666.0            -66.6             0.17           -666.0   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44           -666.0            -66.6   
1            -66.6             0.44           -666.0            -66.6   
2            -66.6             0.45           -666.0            -66.6   
3            -66.6             0.45           -666.0            -66.6   
4            -66.6             0.46           -666.0            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                    137.0                     81.0                     31.0   
1                    138.0                     83.0                     31.0   
2                    139.0                     84.0                     31.0   
3                    140.0                     85.0                     32.0   
4                    141.0                     86.0                     32.0   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                     11.0                      4.0                      0.0   
1                     11.0                      4.0                      0.0   
2                     11.0                      4.0                      0.0   
3                     11.0                      4.0                      0.0   
4                     11.0                      4.0                      0.0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                      3.0                      3.0                      9.0   
1                      3.0                      3.0                     10.0   
2                      3.0                      3.0                     10.0   
3                      3.0                      2.0                     10.0   
4                      3.0                      2.0                     10.0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                      6.0                      0.0                  -9999.0   
1                      6.0                      0.0                  -9999.0   
2                      6.0                      0.0                  -9999.0   
3                      6.0                      0.0                  -9999.0   
4                      6.0                      0.0                  -9999.0   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION LATITUDE  \
0                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
1                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
2                  -9999.0  MITCHELL CAVERNS CA US    1325.9  34.9436   
3                  -9999.0  M

In [13]:
outer_merged.shape

(127020, 48)

Si recuerda que cuando verificó el atributo `Shape` de `climate_temp`, verá que el número de filas en `outer_Merged` coincide con eso. Con una unión externa, puede esperar tener el mismo número de filas que el marco de datos más grande, ya que ninguno se pierde como lo están en una unión interna.

### Left Join
También conocido como una unión externa izquierda. En esta unión, conservará filas que no tienen coincidencias solo en el marco de datos izquierdo (o primero) para fusionar.

In [14]:
left_merged = pd.merge(
    climate_temp, precip_one_station, how="left", on=["STATION", "DATE"]
)
left_merged.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

  STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0            NaN              NaN              NaN              NaN   
1            NaN              NaN              NaN              NaN   
2            NaN              NaN              NaN              NaN   
3            NaN              NaN              NaN              NaN   
4            NaN              NaN              NaN              NaN   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0              NaN              NaN                      NaN   
1              NaN              NaN                      NaN   
2              NaN              NaN                      NaN   
3              NaN              NaN                      NaN   
4              NaN              NaN                      NaN   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
0                      NaN                      

In [15]:
left_merged.shape

(127020, 48)

Aquí, ves que el número de filas en el marco de datos resultante coincide con el de las filas en el `climat_temp` DataFrame. ¿Qué pasa si cambiamos las posiciones de los dos marcos de datos que estamos fusionando?

In [16]:
left_merged_reversed = pd.merge(
    precip_one_station, climate_temp, how="left", on=["STATION", "DATE"]
)
left_merged_reversed.head()

STATION          STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION LATITUDE  \
0                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
1                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
2                    -9999  MITCHELL CAVERNS CA US    1325.9  34.9436   
3                    -9999  M

In [17]:
left_merged_reversed.shape

(365, 48)

### Right Join
Esto funciona igual que la unión a la izquierda, sin embargo, las filas no coincidentes solo se conservan en el _right_ dataFrame. En el siguiente ejemplo, recreará el marco de datos `left_merged` pero con una unión derecha.

In [18]:
right_merged = pd.merge(
    precip_one_station, climate_temp, how="right", on=["STATION", "DATE"]
)
right_merged.head()

STATION STATION_NAME_x      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099            NaN  20100101              NaN   
1  GHCND:USC00049099            NaN  20100102              NaN   
2  GHCND:USC00049099            NaN  20100103              NaN   
3  GHCND:USC00049099            NaN  20100104              NaN   
4  GHCND:USC00049099            NaN  20100105              NaN   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0              NaN              NaN              NaN              NaN   
1              NaN              NaN              NaN              NaN   
2              NaN              NaN              NaN              NaN   
3              NaN              NaN              NaN              NaN   
4              NaN              NaN              NaN              NaN   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                      NaN                      NaN                      NaN   
1                      NaN                      NaN                      NaN   
2                      NaN                      NaN                      NaN   
3                      NaN                      NaN                      NaN   
4                      NaN                      NaN                      NaN   

   DLY-SNOW-PCTALL-GE050TI          STATION_NAME_y ELEVATION  LATITUDE  \
0                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
1                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
2                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
3                      NaN  TWENTYNINE PALMS CA US       602  34.12806   
4    

In [19]:
right_merged.shape

(127020, 48)

Aquí, simplemente volteó las posiciones de los marcos de datos de entrada y especificó una unión derecha. Cuando inspeccione `right_merged`, puede notar que no es exactamente lo mismo que `left_merged`. La única diferencia entre las dos es el orden de las columnas: las columnas de la primera entrada siempre serán las primeras en el marco de datos recién formado.

## .join()
`.join()` usa `merge()` debajo del capó, pero proporciona una interfaz mucho más simplificada para `merge()` y por defecto se une a los índices. Aquí hay un ejemplo introductorio utilizando los parámetros `lsuffix` y `rsuffix` para manejar los nombres de columnas superpuestas.

In [20]:
precip_one_station.join(climate_temp, lsuffix="_left", rsuffix="_right")

STATION_left       STATION_NAME_left  DATE_left  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US   20100105             0.05   
...                 ...                     ...        ...              ...   
1820  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101227             0.04   
1821  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101228             0.04   
1822  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101229             0.04   
1823  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101230             0.04   
1824  GHCND:USC00045721  MITCHELL CAVERNS CA US   20101231             0.04   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   
...               ...              ...              ...              ...   
1820             -666            -66.6             0.15             -666   
1821             -666            -66.6             0.15             -666   
1822             -666            -66.6             0.15             -666   
1823             -666            -66.6             0.15             -666   
1824             -666            -66.6             0.15             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   
...               ...              ...              ...              ...   
1820            -66.6             0.44             -666            -66.6   
1821            -66.6             0.43             -666            -66.6   
1822            -66.6             0.43             -666            -66.6   
1823            -66.6             0.43             -666            -66.6   
1824            -66.6             0.44             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   
...               ...              ...              ...              ...   
1820             1.14              0.3            11.29              1.5   
1821             1.19              0.3            11.34              1.5   
1822             1.24              0.3            11.39              1.5   
1823             1.30              0.3            11.45              1.5   
1824             1.35              0.3            11.50              1.5   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141    

Si inspecciona los datos, verá que las columnas superpuestas se mantienen, solo se renombrará para ser únicas. Si volteamos esto y, en su lugar, llamamos `.join()` En el mejor marcado de datos, notará que el marco de datos es más grande, pero los datos que no existen en el cuadro de datos más pequeño (`precip_one_station`) se llena con `Nan` (_not a Number_).

In [21]:
climate_temp.join(precip_one_station, lsuffix="_left", rsuffix="_right")

STATION_left       STATION_NAME_left ELEVATION  LATITUDE  \
0       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
1       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
2       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
3       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
4       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
...                   ...                     ...       ...       ...   
127015  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127016  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127017  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127018  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127019  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   

         LONGITUDE  DATE_left  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
0       -116.03694   20100101                6                2   
1       -116.03694   20100102                6                2   
2       -116.03694   20100103                6                2   
3       -116.03694   20100104                6                2   
4       -116.03694   20100105                6                2   
...            ...        ...              ...              ...   
127015   -118.0711   20101227                4                2   
127016   -118.0711   20101228                4                2   
127017   -118.0711   20101229                4                2   
127018   -118.0711   20101230                4                2   
127019   -118.0711   20101231                4                2   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
0                 -7777            -7777            -7777                0   
1                     1            -7777            -7777                0   
2                     1            -7777            -7777                0   
3                     1            -7777            -7777                0   
4                     1            -7777            -7777                0   
...                 ...              ...              ...              ...   
127015                1            -7777            -7777            -7777   
127016                1            -7777            -7777            -7777   
127017                1            -7777            -7777            -7777   
127018                1            -7777            -7777            -7777   
127019                1            -7777            -7777            -7777   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
0                     0                0            -7777                1   
1                     0                0            -7777                1   
2                     0                0            -7777                1   
3                     0                0            -7777                1   
4                     0                0            -7777            -7777   
...                 ...              ...              ...              ...   
127015                0                0                2                4   
127016                0                0                2                3   
127017                0                0                2                3   
127018                0                0                2                3   
127019                0                0                1                3   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
0                     2                6                7               10   
1                     2                6                7               10   
2                     2                5                7               10   
3                     2                5                7               10   
4                     2                5                7          

Si debe usar `.join()` y desea fusionar las columnas, debe establecerlas para que sean índices primero. Primero eche un vistazo a esta operación `merge()` utilizada anteriormente:

In [22]:
inner_merged_total = pd.merge(
    climate_temp, climate_precip, on=["STATION", "DATE"]
)
inner_merged_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

In [23]:
inner_joined_total = climate_temp.join(
    climate_precip.set_index(["STATION", "DATE"]),
    on=["STATION", "DATE"],
    how="inner",
    lsuffix="_x",
    rsuffix="_y",
)
inner_joined_total.head()

STATION          STATION_NAME_x ELEVATION  LATITUDE   LONGITUDE  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806  -116.03694   

       DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  DLY-CLDD-BASE55  \
0  20100101                6                2            -7777   
1  20100102                6                2                1   
2  20100103                6                2                1   
3  20100104                6                2                1   
4  20100105                6                2                1   

   DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  DLY-CLDD-BASE70  \
0            -7777            -7777                0                0   
1            -7777            -7777                0                0   
2            -7777            -7777                0                0   
3            -7777            -7777                0                0   
4            -7777            -7777                0                0   

   DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
0                0            -7777                1                2   
1                0            -7777                1                2   
2                0            -7777                1                2   
3                0            -7777                1                2   
4                0            -7777            -7777                2   

   DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  DLY-HTDD-NORMAL  \
0                6                7               10               15   
1                6                7               10               15   
2                5                7               10               15   
3                5                7               10               15   
4                5                7               10               15   

           STATION_NAME_y  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
0  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
1  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
2  TWENTYNINE PALMS CA US            -6.66             -666            -66.6   
3  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   
4  TWENTYNINE PALMS CA US            -6.66            -9999          -9999.0   

   DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
0            -6.66             -666            -66.6            -6.66   
1            -6.66             -666            -66.6            -6.66   
2            -6.66             -666            -66.6            -6.66   
3            -6.66            -9999          -9999.0            -6.66   
4            -6.66            -9999          -9999.0            -6.66   

   DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
0             -666            -66.6             0.02              0.0   
1             -666            -66.6             0.04              0.0   
2             -666            -66.6             0.05              0.0   
3            -9999          -9999.0             0.07              0.0   
4            -9999          -9999.0             0.09              0.0   

   YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
0             0.02              0.0                       98   
1             0.04              0.0                       99   
2             0.05              0.0                      100   
3             0.07              0.0                      101   
4             0.09              0.0                      102   

   DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100

Porque `.join()` funciona en índices, si queremos recrear `merge()` antes, entonces debemos establecer índices en las columnas de unión que especificamos. En este ejemplo, utilizó el método `.set_index()` para establecer sus índices en las columnas clave dentro de la unión.

A continuación, verá una llamada `.join()`. Debido a que hay columnas superpuestas, tendrá que especificar un sufijo con `lsuffix`, `rsuffix` o ambos, pero este ejemplo demostrará el comportamiento más típico de `.join()`.

In [24]:
climate_temp.join(climate_precip, lsuffix="_left")

STATION_left       STATION_NAME_left ELEVATION  LATITUDE  \
0       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
1       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
2       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
3       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
4       GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
...                   ...                     ...       ...       ...   
127015  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127016  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127017  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127018  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   
127019  GHCND:USC00046006  MOUNT WILSON CBS CA US    1740.4   34.2308   

         LONGITUDE  DATE_left  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
0       -116.03694   20100101                6                2   
1       -116.03694   20100102                6                2   
2       -116.03694   20100103                6                2   
3       -116.03694   20100104                6                2   
4       -116.03694   20100105                6                2   
...            ...        ...              ...              ...   
127015   -118.0711   20101227                4                2   
127016   -118.0711   20101228                4                2   
127017   -118.0711   20101229                4                2   
127018   -118.0711   20101230                4                2   
127019   -118.0711   20101231                4                2   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
0                 -7777            -7777            -7777                0   
1                     1            -7777            -7777                0   
2                     1            -7777            -7777                0   
3                     1            -7777            -7777                0   
4                     1            -7777            -7777                0   
...                 ...              ...              ...              ...   
127015                1            -7777            -7777            -7777   
127016                1            -7777            -7777            -7777   
127017                1            -7777            -7777            -7777   
127018                1            -7777            -7777            -7777   
127019                1            -7777            -7777            -7777   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
0                     0                0            -7777                1   
1                     0                0            -7777                1   
2                     0                0            -7777                1   
3                     0                0            -7777                1   
4                     0                0            -7777            -7777   
...                 ...              ...              ...              ...   
127015                0                0                2                4   
127016                0                0                2                3   
127017                0                0                2                3   
127018                0                0                2                3   
127019                0                0                1                3   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
0                     2                6                7               10   
1                     2                6                7               10   
2                     2                5                7               10   
3                     2                5                7               10   
4                     2                5                7          

## concat()

Primero, verá una concatenación básica a lo largo del eje 0.

In [25]:
double_precip = pd.concat([precip_one_station, precip_one_station])
double_precip.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
1460  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1461  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
1462  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
1463  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
1464  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

      DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
1460             -666            -66.6             0.16             -666   
1461             -666            -66.6             0.16             -666   
1462             -666            -66.6             0.16             -666   
1463             -666            -66.6             0.16             -666   
1464             -666            -66.6             0.17             -666   

      DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
1460            -66.6             0.44             -666            -66.6   
1461            -66.6             0.44             -666            -66.6   
1462            -66.6             0.45             -666            -66.6   
1463            -66.6             0.45             -666            -66.6   
1464            -66.6             0.46             -666            -66.6   

      MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
1460             0.04              0.0             0.04              0.0   
1461             0.08              0.0             0.08              0.0   
1462             0.12              0.0             0.12              0.0   
1463             0.16              0.0             0.16              0.0   
1464             0.21              0.0             0.21              0.0   

      DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  \
1460                      137                       81   
1461                      138                       83   
1462                      139                       84   
1463                      140                       85   
1464                      141                       86   

      DLY-PRCP-PCTALL-GE050HI  DLY-PRCP-PCTALL-GE100HI  \
1460                       31                       11   
1461                       31                       11   
1462                       31                       11   
1463                       32                       11   
1464                       32                       11   

      DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
1460                        4                        0   
1461                        4                        0   
1462                        4                        0   
1463                        4                        0   
1464                        4                        0   

      DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  \
1460                        3                        3   
1461                        3                        3   
1462                        3                        3   
1463                        3                        2   
1464                        3                        2   

      DLY-SNOW-PCTALL-GE001TI  DLY-SNOW-PCTALL-GE010TI  \
1460                        9                        6   
1461                       10                        6   
1462                       10                        6   
1463                       10                        6   
1464                       10                        6   

      DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
1460                        0                    -9999   
1461                        0                    -9999   
1462                        0                    -9999   
1463                        0                    -9999   
1464                        0                    -9999   

      DLY-SNOW-PCTALL-GE050TI  
1460                    -9999  
1461              

Para restablecer el índice, use el parámetro opcional `ignore_index`.

In [26]:
reindexed = pd.concat(
    [precip_one_station, precip_one_station], ignore_index=True
)
reindexed.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100101             0.04   
1  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100102             0.05   
2  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100103             0.05   
3  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100104             0.05   
4  GHCND:USC00045721  MITCHELL CAVERNS CA US  20100105             0.05   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0             -666            -66.6             0.16             -666   
1             -666            -66.6             0.16             -666   
2             -666            -66.6             0.16             -666   
3             -666            -66.6             0.16             -666   
4             -666            -66.6             0.17             -666   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6             0.44             -666            -66.6   
1            -66.6             0.44             -666            -66.6   
2            -66.6             0.45             -666            -66.6   
3            -66.6             0.45             -666            -66.6   
4            -66.6             0.46             -666            -66.6   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.04              0.0             0.04              0.0   
1             0.08              0.0             0.08              0.0   
2             0.12              0.0             0.12              0.0   
3             0.16              0.0             0.16              0.0   
4             0.21              0.0             0.21              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                      137                       81                       31   
1                      138                       83                       31   
2                      139                       84                       31   
3                      140                       85                       32   
4                      141                       86                       32   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                       11                        4                        0   
1                       11                        4                        0   
2                       11                        4                        0   
3                       11                        4                        0   
4                       11                        4                        0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                        3                        3                        9   
1                        3                        3                       10   
2                        3                        3                       10   
3                        3                        2                       10   
4                        3                        2                       10   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                        6                        0                    -9999   
1                        6                        0                    -9999   
2                        6                        0                    -9999   
3                        6                        0                    -9999   
4                        6                        0                    -9999   

   DLY-SNOW-PCTALL-GE050TI  
0                    -9999  
1                    -9999  
2                    -9999  
3                    -9999  
4                    -9999

Cuando las etiquetas del eje para el eje, si **no** esta concatenando a lo largo de no coincidir (por ejemplo, las etiquetas de columna al concatenar a lo largo de las filas), entonces todas las columnas se conservan y los datos faltantes se llenan con `nan`.

In [27]:
outer_joined = pd.concat([climate_precip, climate_temp])
outer_joined.head()

STATION            STATION_NAME      DATE  DLY-PRCP-25PCTL  \
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101            -6.66   
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102            -6.66   
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103            -6.66   
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104            -6.66   
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105            -6.66   

   DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
0           -666.0            -66.6            -6.66           -666.0   
1           -666.0            -66.6            -6.66           -666.0   
2           -666.0            -66.6            -6.66           -666.0   
3          -9999.0          -9999.0            -6.66          -9999.0   
4          -9999.0          -9999.0            -6.66          -9999.0   

   DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  \
0            -66.6            -6.66           -666.0            -66.6   
1            -66.6            -6.66           -666.0            -66.6   
2            -66.6            -6.66           -666.0            -66.6   
3          -9999.0            -6.66          -9999.0          -9999.0   
4          -9999.0            -6.66          -9999.0          -9999.0   

   MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  \
0             0.02              0.0             0.02              0.0   
1             0.04              0.0             0.04              0.0   
2             0.05              0.0             0.05              0.0   
3             0.07              0.0             0.07              0.0   
4             0.09              0.0             0.09              0.0   

   DLY-PRCP-PCTALL-GE001HI  DLY-PRCP-PCTALL-GE010HI  DLY-PRCP-PCTALL-GE050HI  \
0                     98.0                     43.0                     12.0   
1                     99.0                     44.0                     12.0   
2                    100.0                     44.0                     12.0   
3                    101.0                     45.0                     12.0   
4                    102.0                     45.0                     12.0   

   DLY-PRCP-PCTALL-GE100HI  DLY-SNWD-PCTALL-GE001WI  DLY-SNWD-PCTALL-GE010WI  \
0                      3.0                  -9999.0                      0.0   
1                      3.0                  -9999.0                      0.0   
2                      3.0                  -9999.0                      0.0   
3                      3.0                      0.0                      0.0   
4                      3.0                      0.0                      0.0   

   DLY-SNWD-PCTALL-GE003WI  DLY-SNWD-PCTALL-GE005WI  DLY-SNOW-PCTALL-GE001TI  \
0                  -9999.0                  -9999.0                  -9999.0   
1                  -9999.0                  -9999.0                  -9999.0   
2                  -9999.0                  -9999.0                  -9999.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   DLY-SNOW-PCTALL-GE010TI  DLY-SNOW-PCTALL-GE100TI  DLY-SNOW-PCTALL-GE030TI  \
0                  -9999.0                      0.0                  -9999.0   
1                  -9999.0                      0.0                  -9999.0   
2                  -9999.0                      0.0                  -9999.0   
3                      0.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   DLY-SNOW-PCTALL-GE050TI ELEVATION LATITUDE LONGITUDE  DLY-CLDD-BASE45  \
0                  -9999.0       NaN      NaN       NaN              NaN   
1                  -9999.0       NaN      NaN       NaN              NaN   
2                  -9999.0       NaN      NaN       NaN              NaN   
3                

In [28]:
inner_joined = pd.concat([climate_temp, climate_precip], join="inner")
inner_joined.head()

STATION            STATION_NAME      DATE
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105

In [29]:
inner_joined.shape

(278130, 3)

Para ilustrar cómo funcionaría esto con filas, concatenate a lo largo de las columnas en su lugar:

In [30]:
inner_joined_cols = pd.concat(
    [climate_temp, climate_precip], axis="columns", join="inner"
)
inner_joined.head()

STATION            STATION_NAME      DATE
0  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100101
1  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100102
2  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100103
3  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100104
4  GHCND:USC00049099  TWENTYNINE PALMS CA US  20100105

In [31]:
inner_joined_cols.shape

(127020, 50)

También puede usar el parámetro `keys` para establecer etiquetas del eje jerárquico que se pueden usar, por ejemplo, para preservar las etiquetas originales mientras mantienen etiquetas que le indican de qué conjunto de datos provienen cada fila o columna.

In [32]:
hierarchical_keys = pd.concat(
    [climate_temp, climate_precip], keys=["temp", "precip"]
)
hierarchical_keys.head()

STATION            STATION_NAME ELEVATION  LATITUDE  \
temp 0  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     1  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     2  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     3  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   
     4  GHCND:USC00049099  TWENTYNINE PALMS CA US       602  34.12806   

         LONGITUDE      DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
temp 0  -116.03694  20100101              6.0              2.0   
     1  -116.03694  20100102              6.0              2.0   
     2  -116.03694  20100103              6.0              2.0   
     3  -116.03694  20100104              6.0              2.0   
     4  -116.03694  20100105              6.0              2.0   

        DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  DLY-CLDD-NORMAL  \
temp 0          -7777.0          -7777.0          -7777.0              0.0   
     1              1.0          -7777.0          -7777.0              0.0   
     2              1.0          -7777.0          -7777.0              0.0   
     3              1.0          -7777.0          -7777.0              0.0   
     4              1.0          -7777.0          -7777.0              0.0   

        DLY-CLDD-BASE70  DLY-CLDD-BASE72  DLY-HTDD-BASE40  DLY-HTDD-BASE45  \
temp 0              0.0              0.0          -7777.0              1.0   
     1              0.0              0.0          -7777.0              1.0   
     2              0.0              0.0          -7777.0              1.0   
     3              0.0              0.0          -7777.0              1.0   
     4              0.0              0.0          -7777.0          -7777.0   

        DLY-HTDD-BASE50  DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
temp 0              2.0              6.0              7.0             10.0   
     1              2.0              6.0              7.0             10.0   
     2              2.0              5.0              7.0             10.0   
     3              2.0              5.0              7.0             10.0   
     4              2.0              5.0              7.0             10.0   

        DLY-HTDD-NORMAL  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  DLY-SNOW-25PCTL  \
temp 0             15.0              NaN              NaN              NaN   
     1             15.0              NaN              NaN              NaN   
     2             15.0              NaN              NaN              NaN   
     3             15.0              NaN              NaN              NaN   
     4             15.0              NaN              NaN              NaN   

        DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  \
temp 0              NaN              NaN              NaN              NaN   
     1              NaN              NaN              NaN              NaN   
     2              NaN              NaN              NaN              NaN   
     3              NaN              NaN              NaN              NaN   
     4              NaN              NaN              NaN              NaN   

        DLY-SNWD-75PCTL  DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
temp 0              NaN              NaN              NaN              NaN   
     1              NaN              NaN              NaN              NaN   
     2              NaN              NaN              NaN              NaN   
     3              NaN              NaN              NaN              NaN   
     4              NaN              NaN              NaN              NaN   

        YTD-PRCP-NORMAL  YTD-SNOW-NORMAL  DLY-PRCP-PCTALL-GE001HI  \
temp 0              NaN              NaN                      NaN   
     1              NaN              NaN                      NaN   
     2              NaN              NaN                      NaN   
     3              NaN              NaN                      NaN   
     4              NaN

In [33]:
hierarchical_keys.tail()

STATION            STATION_NAME ELEVATION LATITUDE  \
precip 151105  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151106  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151107  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151108  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   
       151109  GHCND:USC00046006  MOUNT WILSON CBS CA US       NaN      NaN   

              LONGITUDE      DATE  DLY-CLDD-BASE45  DLY-CLDD-BASE50  \
precip 151105       NaN  20101227              NaN              NaN   
       151106       NaN  20101228              NaN              NaN   
       151107       NaN  20101229              NaN              NaN   
       151108       NaN  20101230              NaN              NaN   
       151109       NaN  20101231              NaN              NaN   

               DLY-CLDD-BASE55  DLY-CLDD-BASE57  DLY-CLDD-BASE60  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-CLDD-NORMAL  DLY-CLDD-BASE70  DLY-CLDD-BASE72  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-BASE40  DLY-HTDD-BASE45  DLY-HTDD-BASE50  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-BASE55  DLY-HTDD-BASE57  DLY-HTDD-BASE60  \
precip 151105              NaN              NaN              NaN   
       151106              NaN              NaN              NaN   
       151107              NaN              NaN              NaN   
       151108              NaN              NaN              NaN   
       151109              NaN              NaN              NaN   

               DLY-HTDD-NORMAL  DLY-PRCP-25PCTL  DLY-SNWD-25PCTL  \
precip 151105              NaN             0.12          -9999.0   
       151106              NaN             0.12          -9999.0   
       151107              NaN             0.12          -9999.0   
       151108              NaN             0.12          -9999.0   
       151109              NaN             0.12          -9999.0   

               DLY-SNOW-25PCTL  DLY-PRCP-50PCTL  DLY-SNWD-50PCTL  \
precip 151105          -9999.0             0.41          -9999.0   
       151106          -9999.0             0.42          -9999.0   
       151107          -9999.0             0.43          -9999.0   
       151108          -9999.0             0.43          -9999.0   
       151109          -9999.0             0.44          -9999.0   

               DLY-SNOW-50PCTL  DLY-PRCP-75PCTL  DLY-SNWD-75PCTL  \
precip 151105          -9999.0             1.13          -9999.0   
       151106          -9999.0             1.14          -9999.0   
       151107          -9999.0             1.15          -9999.0   
       151108          -9999.0             1.15          -9999.0   
       151109          -9999.0             1.16          -9999.0   

               DLY-SNOW-75PCTL  MTD-PRCP-NORMAL  MTD-SNOW-NORMAL  \
precip 151105          -9999.0             3.91          -9999.0   
       151106          -9999.0             4.10          -9999.0   
       151107          -9999.0             4.30         